python/mxnet/optimizer.py

```diff
@@ -750,12 +754,39 @@ def update(self, index, weight, grad, state):
         lr = self._get_lr(index)
         wd = self._get_wd(index)
 
+        is_sparse = True if weight.stype == 'row_sparse' and grad.stype == 'row_sparse' else False
+
+        if is_sparse is True:
```

Comment:
Really? Does scipy do that as well?

Reply:
Yes.. 
@piiswrong, @reminisce and I discussed this before, making a reference causes many memory issues for the ndarray in backend because all aux_data/data are wrapped in a single chunk, and it uses a single engine::variable for the chunk. Wrapping an aux_data as an ndarray is not thread safe.

Reply:
ic